Add Reserved Template Name Sniff

dingo-d opened the pull request:
This is a PR for #212 issue.

I'm sure more test cases would be a good idea, and I'm not 100% sure if I should check for the comment at all, or just check for prefixed file names.

One thing that I had to add was `strip_quotes` method from the WPCS `Sniff` abstract class.

Should I have just extended that sniff instead of implementing the `Sniff` interface from PHPCS (seeing how WPCS's abstract `Sniff` class already implements it).

jrfnl commented:
> One thing that I had to add was strip_quotes method from the WPCS Sniff abstract class.
> 
> Should I have just extended that sniff instead of implementing the Sniff interface from PHPCS (seeing how WPCS's abstract Sniff class already implements it).

If things go as planned :crossed_fingers: and PR squizlabs/PHP_CodeSniffer#2456 will be accepted, PHPCS 3.5.0 will contain that function natively: https://gist.github.com/jrfnl/16e1d64d9ec8efec7729e9bfbb05201a#in-php_codesnifferutilsniffstextstring

dingo-d commented:
Oh the mega PR 😄 

So should I leave it as is and refactor after that, or extend the WPCS `Sniff` and refactor after that? 😄

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 *
+	 * @var array
+	 */
+	protected $reserved_file_name_prefixes = [ 'index-', 'comments-', 'front-page-', 'home-', 'header-', 'singular-', 'single-', 'page-', 'category-', 'tag-', 'taxonomy-', 'author-', 'date-', 'archive-', 'search-', 'attachment-', 'image-', '404-' ];
```
Please turn this into a multi-line array for readability.

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 *
+	 * @var array
+	 */
+	protected $reserved_file_name_prefixes = [ 'index-', 'comments-', 'front-page-', 'home-', 'header-', 'singular-', 'single-', 'page-', 'category-', 'tag-', 'taxonomy-', 'author-', 'date-', 'archive-', 'search-', 'attachment-', 'image-', '404-' ];
```
If I read the theme handbook, the template hierarchy and the WP Core code for loading templates correctly, the only ones which would be problematic are:
* `author-`
* `category-`
* `tag-`
* `page-`

The problem occurs if the WP logic for including a template could confuse the file name for `type-$slug` and AFAICS, those are the only ones where that could happen.

A template name like `comments-something.php` would just be ignored by WP and have no effect, except when specifically `include`d by theme from one of the other files.

dingo-d replied:
Yes, I thought about this too, but didn't find any additional information about if people are doing such thing (like `index-secondary.php` or `comments-lists.php`).

But thought - well it doesn't hurt to catch these cases as well 🤷‍♂Maybe we could warn the users if they use something like `index-secondary.php`, `home-secondary.php` or `comments-something.php` that WordPress will ignore these. Although the same would happen with the above due to how the WP template hierarchy works, so I guess it makes no sense to split these two.

The biggest issue for me is that I don't have the concrete number or evidence if these sort of things happen at all (the index prefix for instance). I'll try to get all the themes and search through them to see if I'm getting any positives on this

jrfnl replied:
> So should we check for all the prefixes?

The question is: what is this sniff about ?
Is it about templates which will be interpreted as `$slug` specific by WP ? If so, then no, this should not be in this sniff or at the very least, it should have a different error code.
Those other templates will be ignored by WP natively, but can be included from within the theme from another template file, checking whether that is done is a whole other matter.

dingo-d replied:
I see. Well the original issue was with `$slug` specific so only 4 problematic ones are the ones you mentioned :+1:

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Check if the template file is using reserved file name as a part of their name.
```
```suggestion
 * Check if the template file is using a prefix which would cause WP to interpret it as a specialised template
 * meant to apply to only one page on the site.
```

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 *
+	 * @var string
+	 */
+	const ERROR_MSG = 'File template should not use a reserved name as a prefix. Found: "%s".';
```
From a dev-education point of view, I think this message could be improved as it doesn't explain _why_.

dingo-d replied:
Would this be a better explanation

```
'Template file should not use a reserved name as a prefix, because it will be recognised by WordPress and will apply to one specific page only. Found: "%s".'
```

?

jrfnl replied:
Still not sure if that's clear enough. Maybe ask some theme reviewers for some suggestions ?

Rough draft from my end:
`Template files should not be slug specific. The file name used for this template will be interpreted by WP as %s-$slug and only applied when the %s named $slug will be loaded.`

dingo-d replied:
I thought it has to be something short 😆 Your suggestion looks great 😄 I'll ask in the channel :+1:

jrfnl replied:
The message can be as long or short. Most importantly, it has to be informative. You can even use new lines in error messages if it helps (though it may annoy people with very wide screens).

Errorcodes should be relatively short and modular, i.e. make it easy for people to ignore just one type of issue.

dingo-d replied:
I've asked and this seems to be better:

`
Template files should not be slug specific. The file name used for this template will be interpreted by WP as %1$s-%2$s and only applied when a %1$s with the slug %2$s is loaded.
`

I'm also capturing the `slug` and `prefix` so the error message would be like this 

```
FILE: /Users/infinum-denis/Desktop/TEST/page-comments.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 1 | ERROR | Template files should not be slug specific. The file
   |       | name used for this template will be interpreted by WP as
   |       | page-comments and only applied when a page with the
   |       | slug comments is loaded.
----------------------------------------------------------------------
```

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 *
+	 * Intended for use with the contents of a T_CONSTANT_ENCAPSED_STRING / T_DOUBLE_QUOTED_STRING.
+	 *
+	 * Used from WordPressCS\WordPress\Sniff abstract class-
```
```suggestion
	 * Copied from the WordPressCS\WordPress\Sniff abstract class.
```

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+
+		// Check if the current file has a prefix in the reserved list.
+		if ( ! $this->is_reserved_template_name_used( $fileName ) ) {
+			return;
```
If the file name does not use one of the reserved prefixes, there is no need to ever examine the file again for this sniff.
```php
return ( $phpcsFile->numTokens + 1 );
```

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 * @param int                         $stackPtr  The position of the current token
+	 *                                               in the stack.
+	 *
+	 * @return void
```
```suggestion
	 * @return int|void ....
```

See https://github.com/squizlabs/PHP_CodeSniffer/blob/1d7c6576a578fb57723c8c0fbc716cb88dae4f13/src/Sniffs/Sniff.php#L72-L75 for inspiration for the return type description.

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 */
+	private function is_reserved_template_name_used( $file ) {
+		foreach ( $this->reserved_file_name_prefixes as $prefix ) {
+			if ( strpos( $file, $prefix ) !== false ) {
```
This would also match `some-other-page-type.php`. Please check for prefixes only.

jrfnl commented on WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,137 @@
+	 * @return boolean
+	 */
+	private function is_reserved_template_name_used( $file ) {
+		foreach ( $this->reserved_file_name_prefixes as $prefix ) {
```
I think this can be made more efficient:
* Get the start of the filename up to the first `-` (if any).
* Change the array of prefixes in the property above to `'prefix' => true` format.
* Just check against the property with `isset ( $this->property[ $prefix ] )`.

No need to loop through the array.

dingo-d replied:
Definitely more performant/efficient way to go about it :+1:

jrfnl replied:
Some notes on the regex:
* If you're not using `$matches[1]`, forgetting/not setting it will be more performant.
* If there's a better alternative than using a `.` in a regex, always use it.
* The regex still doesn't check that the prefix is used at the start of the file name.
* Regexes are heavy and not really needed here.

Regex-wise, this should be changed to `/^[^-]+-/`.

But better yet would be to remove the regex and use something along the lines of:
```php
$dash_position = strpos( $file, '-' );
if ( $dash_position === false || $dash_position === 0 ) {
    // No dash or dash at start of filename, i.e. `-something.php`
    return false;
}

$partial = substr( $file, 0, $dash_position );
```

dingo-d replied:
This is called optimisation 😄 I'll replace this :+1:

jrfnl commented on WPThemeReview/Tests/ThouShallNotUse/ReservedTemplateNameUnitTest.inc:
```diff
@@ -0,0 +1,2 @@
+<?php
+// Dummy file. The real test files are in the ReservedTemplateNameTests subfolder.
```
:+1:

jrfnl commented on WPThemeReview/Tests/ThouShallNotUse/ReservedTemplateNameUnitTest.php:
```diff
@@ -0,0 +1,66 @@
+use PHP_CodeSniffer\Tests\Standards\AbstractSniffUnitTest;
+
+/**
+ * Unit test class for the PrefixAllGlobals sniff.
```
copy/paste

dingo-d commented:
I've fixed the issues, and changed the code as suggested, but I won't commit until I get the approval on the comments I've left (so as to not pollute the commit history) 🙂 

I'm not sure we need the util class in the standards, as most of the time we're extending WPCS sniffs (which are extending the WPCS `Sniff` class). Plus if everything goes as planned, the helper should end up in the PHPCS, and then it would be easy to just remove the helper method from the sniff and use it from  the upstream 🙂

jrfnl commented:
Hope I've caught all the comments you wanted a response to. Let me know if I missed any.

dingo-d commented:
First, the regex comment blew my mind, from 13 passes to 4, and then the fact that I didn't need regex in the first place. Never would have thought of that 😄 I'm super grateful for your help, remind me to buy you a beer at WCEU (the least I can do) 😄

dingo-d commented:
@jrfnl This is also ready for a final review 🙂

jrfnl commented on WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,156 @@
+	 * @param  string $file File name to check.
+	 * @return boolean
+	 */
+	private function is_reserved_template_name_used( $file ) {
```
In the previous iteration of this method, you only looked at the part before the `-`, as you now also want to access/remember the part of the file name after the `-`, there is an even simpler way of doing this: using `explode()` with a limit.

If - like suggested above -, you strip off the `.php` from the filename before passing it to this function, the net result will be already exactly what you need. See the below code snippet (untested, but should work).

```php
$parts = explode( '-', $file, 2 );
if ( empty( $parts[0] ) || empty( $parts[2] ) ) {
	// No dash or dash at start or end of filename, i.e. `-something.php`.
	return false;
}

// $parts[0] will now contain what you call `$partial` in the original code,
// while `$parts[1]` will contain the rest of the file name.
```

dingo-d replied:
I'll change the `$reserved_file_name_prefixes` to include prefixes without the dash. This will solve the [above comment](https://github.com/WPTRT/WPThemeReview/pull/213#discussion_r292825884) as well 🙂

jrfnl commented on WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,156 @@
+	 *
+	 * @var string
+	 */
+	const ERROR_MSG = 'Template files should not be slug specific. The file name used for this template will be interpreted by WP as %1$s-%2$s and only applied when a %1$s with the slug %2$s is loaded.';
```
```suggestion
	const ERROR_MSG = 'Template files should not be slug specific. The file name used for this template will be interpreted by WP as %1$s-%2$s and only applied when a %1$s with the slug "%2$s" is loaded.';
```

jrfnl commented on WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,156 @@
+
+	/**
+	 * Processes this test, when one of its tokens is encountered.
+	 *
```
Missing `@since` tag.

jrfnl commented on WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,156 @@
+
+	/**
+	 * Checks if the given file is located in the $reserved_file_name_prefixes array.
+	 *
```
Missing `@since` tag.

jrfnl commented on WPThemeReview/Tests/Templates/ReservedTemplateNameTests/page-contact.inc:
```diff
@@ -0,0 +1,4 @@
+<?php
+/* Template Name: Contact page template */
+
+// code goes here
```
Might be an idea to have some dummy inline HTML with embedded PHP here to safe-guard that only one error will be thrown per file.

jrfnl commented on WPThemeReview/Tests/Templates/ReservedTemplateNameTests/category-books.inc:
```diff
@@ -0,0 +1,4 @@
+<?php
+/* Template Name: Books category template */
+
+// code goes here
```
Might be an idea to not have a full PHP open tag in this file, but just a short echo open tag `<?=` somewhere in embedded PHP to verify that the sniff also works correctly when encountering templates without proper docblocks.

dingo-d replied:
Not sure I understand this. I need to have a `/* Template Name: Books category template */` in order for WP to recognise the custom templates, no?

```php
<?php
/* Template Name: Books category template */
?>

<div class="some-class-name">
  <?= 'This is an echo'; ?>
</div>
```

This now throws error about the template and the short echo used 😄

jrfnl replied:
From https://github.com/WPTRT/WPThemeReview/pull/213#pullrequestreview-236358438:

> > I'm not 100% sure if I should check for the comment at all, or just check for prefixed file names.
>
> I don't think you should as I don't think WP core does when including the file, so checking for the comments would lead to false negatives. (should be verified though, I haven't dug that deep)

In other words, it needs to be checked if WP Core checks the comment at the top of template files.
If it does, so should this sniff.
If it doesn't, this sniff shouldn't either.

jrfnl replied:
> This now throws error about the template and the short echo used

Correct. Sniffs should work independently of each other.

dingo-d replied:
So my example is correct?

jrfnl replied:
Unfortunately, no.

The sniff will only look at the file once, so with your code sample would get triggered on the first `<?php` and wouldn't get triggered anymore on the `<?=`, so the sniff being triggered on short open echo tags is still not being tested.

dingo-d replied:
So I should put the HTML part before the template comment? I've never seen such a file in a template 😕

jrfnl replied:
> So I should put the HTML part before the template comment? I've never seen such a file in a template

No, the test file I'm suggesting shouldn't have a (PHP) comment, just template code.
Something like:
```html
<!-- Template Name: Books category template -->
<div class="some-class-name">
  <?= 'This is an echo'; ?>
</div>
```

dingo-d commented on WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,156 @@
+ *
+ * @since 0.2.0
+ */
+class ReservedTemplateNameSniff implements Sniff {
```
`ReservedFileNamePrefixSniff` sounds good, I'll change this :+1:

dingo-d commented on WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php:
```diff
@@ -0,0 +1,156 @@
+			return false;
+		}
+
+		if ( isset( $this->reserved_file_name_prefixes[ "$partial-" ] ) ) {
```
That's because the `$partial` will return just the part without the dash, and I need to check if the found partial is in the `$reserved_file_name_prefixes` array which contains dashes, so this
`$this->reserved_file_name_prefixes[ "$partial-" ]` would not match anything because the array looks like this

```php
$reserved_file_name_prefixes = [
	'author-'   => true,
	'category-' => true,
	'page-'     => true,
	'tag-'      => true,
];
```

I could change the `$partial` to 

```php
substr( $file, 0, $dash_position + 1 );
```

And then use it to check, that could be better, right?

jrfnl replied:
Sorry, I think you missed my point - why do the array keys in the `$reserved_file_name_prefixes` all have a dash at the end ?

dingo-d replied:
Yes I changed that. When I read other comments I was like: why are there dashes in that array in the first place? 😄

dingo-d commented:
I've added your suggestion, is all ok now? 🙂

jrfnl commented on WPThemeReview/Sniffs/Templates/ReservedFileNamePrefixSniff.php:
```diff
@@ -0,0 +1,157 @@
+
+	/**
+	 * Error message template.
+	 *
```
Missing `@since`...

jrfnl commented on WPThemeReview/Sniffs/Templates/ReservedFileNamePrefixSniff.php:
```diff
@@ -0,0 +1,157 @@
+			return false;
+		}
+
+		if ( isset( $this->reserved_file_name_prefixes[ $file_parts[0] ] ) ) {
```
Should this account for potentially mixed case template file names ?
```suggestion
		if ( isset( $this->reserved_file_name_prefixes[ strtolower( $file_parts[0] ) ] ) ) {
```

dingo-d commented:
@jrfnl From what I gathered, the `template-loader.php` which is responsible for loading the templates in WordPress, has a series of if/elseif checks that will determine if certain queried object is a category, 404 page, search page etc.

For instance the `get_category_template()` function looks like this

```php
function get_category_template() {
	$category = get_queried_object();

	$templates = array();

	if ( ! empty( $category->slug ) ) {

		$slug_decoded = urldecode( $category->slug );
		if ( $slug_decoded !== $category->slug ) {
			$templates[] = "category-{$slug_decoded}.php";
		}

		$templates[] = "category-{$category->slug}.php";
		$templates[] = "category-{$category->term_id}.php";
	}
	$templates[] = 'category.php';

	return get_query_template( 'category', $templates );
}
```

So the answer to the first question is that core expects lowercase filenames for the core templates.
Maybe I could link the [template hierarchy docs](https://developer.wordpress.org/themes/basics/template-hierarchy/) for this (in the `@see` docblock), or do you want me to link the 4 reserved file name prefixes ( `is_page()`, `is_category()`, `is_tag()`, `is_author()`) documentations?

I have added the suggested test files and all is fine 🙂

jrfnl commented:
> So the answer to the first question is that core expects lowercase filenames for the core templates.

Thank you for checking that. In that case, the change I suggested in https://github.com/WPTRT/WPThemeReview/pull/213#discussion_r301039976 wasn't necessary and should be undone and we should possibly have a test case file with mixed case in the file name which shouldn't trigger the sniff, like `Page-something.inc`.

Sorry for the confusion, but I'm very glad you checked as this thread now documents your findings about the check within WP for these files names being case-sensitive. :+1: 

So, aside from this, all looks absolutely perfectly fine and ready for merge.

> Maybe I could link the template hierarchy docs for this (in the @see docblock), or do you want me to link the 4 reserved file name prefixes ( get_page_template(), get_category_template(), get_tag_template(), get_author_template()) documentations?

> I've added the links to the four functions that we are checking for, and the template hierarchy docs. This should be fine?

:+1:

> I have added the suggested test files and all is fine slightly_smiling_face

:+1:

joyously commented:
Does it matter that only some operating systems are case sensitive, when looking for the file?

dingo-d commented:
I don't think this should be an issue, as the core is handling the file loading, so if you have case sensitive filename it won't load it as a template (`Page-info.php` for instance).

joyously commented:
I'm just saying that on a Windows server, looking for 'page-' would match 'Page-', but not on a Linux server.

jrfnl commented:
> I don't think this should be an issue, as the core is handling the file loading, so if you have case sensitive filename it won't load it as a template (Page-info.php for instance).

Are we sure about that ? Has someone tried/tested this with a real theme on a non-Linux system ?

jrfnl commented:
@dingo-d I'm happy to merge this as it is at the moment. Well done! Great work.

If someone has a chance to check out the whole case-sensitivity (or not) of template loading in core at some point on multiple OSes in more detail and the check would turn out to need to be case-**in**sensitive, an issue can be opened to update this PR.

Does that sound reasonable to you ?

dingo-d commented:
On Windows machine even if you name a template `Page-info.php`, WordPress will load it like if you were to name it `page-info.php`, while on my mac the `Page-info.php` will get ignored, and `page-info.php` will not get ignored.

So we need to take this into the account while sniffing, right?

jrfnl commented:
> On Windows machine even if you name a template Page-info.php, WordPress will load it like if you were to name it page-info.php, while on my mac the Page-info.php will get ignored, and page-info.php will not get ignored.
> 
> So we need to take this into the account while sniffing, right?

If that's the case, then yes, we should.

Implementing my [earlier suggestion](https://github.com/WPTRT/WPThemeReview/pull/213#discussion_r301039976) related to this question should take care of that + the `Page-info.inc` test case file will need to be added to the errors list in the unit test file in that case.

jrfnl commented:
I think this whole case-sensitive template file names discussion was very useful to have and it does raise another question: whether there should be rule that template files should always be in lowercase to prevent issues with templates loading on Windows, but not loading on Linux/Mac (and a sniff to verify this).

Not something to be addressed in this sniff, but someone may want to open a new issue about this and this should possibly be discussed in a TRT meeting ?

dingo-d commented:
Squashed all the commits and added the changes you mentioned.

I've brought your comment up with the leads, and I'll mention it on todays meeting if there will be the time because this could be added to the themes handbook which we'll have access to :+1:

jrfnl commented:
Yeah! Merged :tada:

dingo-d commented:
Awesome 😄 Thanks! We're nearing the 0.2.0 release 😄
